service/src/main/kotlin/fi/espoo/evaka/attendance/MobileRealtimeStaffAttendanceController.kt

```diff
@@ -193,6 +198,84 @@ class MobileRealtimeStaffAttendanceController(private val ac: AccessControl) {
+        val date: LocalDate,
+        val rows: List<RealtimeStaffAttendanceController.StaffAttendanceUpsert>
+    )
+
+    data class StaffAttendanceUpdateResponse(
+        val deleted: List<StaffAttendanceRealtimeId>,
+        val updated: List<StaffAttendanceRealtimeId>
+    )
+
+    @PutMapping
+    fun setAttendances(
+        db: Database,
+        user: AuthenticatedUser,
+        clock: EvakaClock,
+        @RequestParam unitId: DaycareId,
+        @RequestBody body: StaffAttendanceUpdateRequest
+    ): StaffAttendanceUpdateResponse {
+        return db.connect { dbc ->
+            dbc.transaction { tx ->
+                ac.requirePermissionFor(
+                    tx,
+                    user,
+                    clock,
+                    Action.Unit.UPDATE_STAFF_ATTENDANCES,
+                    unitId
+                )
+                ac.verifyPinCodeAndThrow(tx, body.employeeId, body.pinCode, clock)
+
+                if (
+                    body.rows.any {
+                        !DateRange(it.arrived.toLocalDate(), it.departed?.toLocalDate())
+                            .includes(body.date)
+                    }
+                ) {
+                    throw BadRequest("Attendances outside given date")
+                }
+                val groupIds = body.rows.mapNotNull { it.groupId }.distinct()
+                if (!groupIds.all { groupId -> tx.getDaycareIdByGroup(groupId) == unitId }) {
+                    throw BadRequest("Group is not in unit")
+                }
+
+                val existing = tx.getEmployeeAttendancesForDate(unitId, body.employeeId, body.date)
+
+                val ids = body.rows.mapNotNull { it.id }
+                if (ids.any { id -> !existing.map { it.id }.contains(id) }) {
+                    throw BadRequest("Unknown id was given")
+                }
+
+                val deletedIds = existing.map { it.id }.filterNot { ids.contains(it) }
+                deletedIds.forEach { tx.deleteStaffAttendance(it) }
+
+                val occupancyCoefficients =
+                    groupIds.associateWith { groupId ->
+                        tx.getOccupancyCoefficientForEmployee(body.employeeId, groupId)
+                            ?: occupancyCoefficientZero
+                    }
+                val updatedIds =
+                    body.rows.map { attendance ->
+                        tx.upsertStaffAttendance(
+                            attendanceId = attendance.id,
+                            employeeId = body.employeeId,
+                            groupId = attendance.groupId,
+                            arrivalTime = attendance.arrived,
+                            departureTime = attendance.departed,
+                            occupancyCoefficient = occupancyCoefficients[attendance.groupId]
+                                    ?: occupancyCoefficientZero,
+                            type = attendance.type
+                        )
+                    }
```

Comment:
It would be better to delete all entries that intersect with `body.date` and then insert fresh rows from `body.rows`. Otherwise the order of upserts will affect whether some changed are allowed.

For example, the initial rows can be:

| id | arrived | departed |
| --- | --- | --- |
| 1 |08:00 |12:00 |
| 2 | 14:00|18:00 |

The user submits:

| id | arrived | departed |
| --- | --- | --- |
| 1 |08:00 |15:00 |
| 2 | 16:00|18:00 |

If the upsert order is id 1, then id 2, the first upsert will conflict with the existing row 2 in the database.

Reply:
Fixed